llvm/lib/Target/EraVM/MCTargetDesc/EraVMMCCodeEmitter.cpp

```diff
@@ -60,15 +66,66 @@ class EraVMMCCodeEmitter : public MCCodeEmitter {
                         SmallVectorImpl<MCFixup> &Fixups,
                         const MCSubtargetInfo &STI) const;
 
+  uint64_t fixOpcode(const MCInst &MI, uint64_t EncodedInstr,
+                     const MCSubtargetInfo &STI) const;
+
   void encodeInstruction(const MCInst &MI, raw_ostream &OS,
                          SmallVectorImpl<MCFixup> &Fixups,
                          const MCSubtargetInfo &STI) const override;
+
+private:
+  uint64_t getRegWithAddend(const MCInst &MI, unsigned BaseReg,
+                            int Addend) const;
 };
 
+int modeEncodingByMarker(const MCOperand &Op) {
+  assert((Op.isImm() || Op.isReg()) && "Unexpected marker operand");
+
+  if (Op.isImm())
+    return EraVM::ModeStackAbs;
+
+  unsigned MarkerReg = Op.getReg();
+  assert(MarkerReg == EraVM::R0 || MarkerReg == EraVM::SP);
+  return (MarkerReg == EraVM::SP) ? EraVM::ModeSpRel : EraVM::ModeSpMod;
+}
+
+uint64_t EraVMMCCodeEmitter::fixOpcode(const MCInst &MI, uint64_t EncodedInstr,
+                                       const MCSubtargetInfo &STI) const {
+  const MCInstrDesc &Desc = MCII.get(MI.getOpcode());
+  auto GetOpIndex = [&Desc](unsigned UseIndex) {
+    return EraVM::getLogicalOperandIdx(Desc.getOpcode(),
```

Comment:
@atrosinenko could you navigate me to the code which produces `getLogicalOperandIdx`, `getLogicalOperandSize`?
The latter is generated to
```
LLVM_READONLY static unsigned
getLogicalOperandSize(uint16_t Opcode, uint16_t LogicalOpIdx) {
   return LogicalOpIdx;
}
```
Which doesn't seem to be correct if I understand the meaning of these functions.

Reply:
@akiramenai This function is tablegen-erated when `let UseLogicalOperandMappings = true` is set in the `Instruction` class (in our case - it is set by this particular patch in the `IForm` class in `EraVMInstrFormats.td`).

To get the correct implementation, you need to build at least once after checking this revision out - but I agree the "default" implementation (and its presence at all) looks strange.